pykeops/numpy/utils.py

```diff
@@ -119,6 +133,81 @@ def array(x, dtype=default_dtype, device=None):
     def device(x):
         return "cpu"
 
+    @staticmethod
+    def distance_function(metric):
+        def euclidean(x, y):
+            return ((x - y) ** 2).sum(-1)
+
+        def manhattan(x, y):
+            return ((x - y).abs()).sum(-1)
+
+        def angular(x, y):
+            return -(x | y)
+
+        def angular_full(x, y):
+            return angular(x, y) / ((angular(x, x) * angular(y, y)).sqrt())
```

Comment:
In order to handle the full angular metric, normalizing the input dataset first should be much more efficient: we'll avoid computing the product of the norms for every single pair of points.

Reply:
We provide the standard option for users to use the shortened angular metric with normalisation.
Alternatively, we also provide an alternative option to use the full angular metric, should they want it.